example/simple/src/main.c

```diff
@@ -81,27 +118,79 @@ int initialize_cache(ocf_ctx_t ctx, ocf_cache_t *cache)
 	if (ret)
 		return ret;
 
+	/*
+	 * Allocate cache private structure. We can not initialize it
+	 * on stack, as it may be used in various async contexts
+	 * throughout the entire live span of cache object.
+	 */
+	cache_priv = malloc(sizeof(*cache_priv));
+	if (!cache_priv)
+		return -ENOMEM;
+
 	/* Start cache */
 	ret = ocf_mngt_cache_start(ctx, cache, &cache_cfg);
 	if (ret)
-		return ret;
+		goto err_priv;
 
-	ret = ocf_queue_create(*cache, &queue, &queue_ops);
-	if (!queue) {
-		ocf_mngt_cache_stop(*cache);
-		return -ENOMEM;
-	}
+	/* Assing cache priv structure to cache. */
+	ocf_cache_set_priv(*cache, cache_priv);
 
-	ocf_cache_set_priv(*cache, queue);
+	/*
+	 * Create management queue. It will be used for performing various
+	 * asynchronous management operations, such as attaching cache volume
+	 * or adding core object.
+	 */
+	ret = ocf_queue_create(*cache, &cache_priv->mngt_queue, &queue_ops);
+	if (ret)
+		goto err_cache;
 
-	/* Attach volume to cache */
-	ret = ocf_mngt_cache_attach(*cache, &device_cfg);
+	/*
+	 * Assign management queue to cache. This has to be done before any
+	 * other management operation. Management queue it treated specially,
+	 * and it may not be used for submitting IO requests. It also will not
+	 * be put on the cache stop - we have to put it manually at the end.
+	 */
+	ocf_mngt_cache_set_mngt_queue(*cache, cache_priv->mngt_queue);
+
+	/* Create queue which will be used for IO submission. */
+	ret = ocf_queue_create(*cache, &cache_priv->io_queue, &queue_ops);
 	if (ret) {
-		ocf_mngt_cache_stop(*cache);
-		return ret;
+		goto err_queue;
 	}
 
+	/* Attach volume to cache */
+	ocf_mngt_cache_attach(*cache, &device_cfg, simple_complete, &context);
+	if (ret)
+		goto err_cache;
```

Comment:
what about putting io_queue in case of err here?

Reply:
The io queue will be put during cache stop anyway, so it would be reduntant.